### PR TITLE
Routing metrics and billing

### PR DESCRIPTION
**Pull request by ethervoid**
**THIS PR [HAS MIGRATION](https://github.com/CartoDB/cartodb/blob/routing_billing/db/migrate/20160212170700_add_routing_quota_and_block_price_fields.rb) SO THE MIGRATION SHOULD BE DEPLOYED BEFORE THIS PR**

This PR has all the need logic to configure and charge the users for the new routing services:

- Metrics for the new routing services
- Overquota logic for routing service
- Refactor to make it easy to add new services
- New field for the organization users to set the soft routing limit
- Tests

### CODE REVIEW

**Comment by ethervoid** [comment]
@rafatower please check it out :)

**Comment by ethervoid** [comment]
@juanignaciosl could you please check this out? thank you!

**Comment by juanignaciosl** [comment]
_Added "do not merge yet" label because of the migration comment, beginning CR..._

File: app/controllers/admin/organization_users_controller.rb [hunk 1]
```diff
@@ -131,13 +132,17 @@ def regenerate_api_key
...
-    extra_geocodings_enabled? || extra_tweets_enabled?
+    extra_geocodings_enabled? || extra_routings_enabled? || extra_tweets_enabled?
   end
 
   def extra_geocodings_enabled?
     !Cartodb.get_config(:geocoder, 'app_id').blank?
   end
 
+  def extra_routings_enabled?
+    true
```
Comment (juanignaciosl) [inline app/controllers/admin/organization_users_controller.rb]: Shouldn't this use `soft_routing_limit` column?

Comment (ethervoid) [inline app/controllers/admin/organization_users_controller.rb]: Take a look to the geocoding part. Is not using the `soft_geocoding_limit` is something strange

File: app/helpers/data_services_metrics_helper.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  private
+
+  def get_geocoding_data(user, from, to)
+    orgname = user.organization.nil? ? nil : user.organization.name
+    usage_metrics = CartoDB::GeocoderUsageMetrics.new(user.username, orgname)
+    geocoder_key = user.google_maps_geocoder_enabled? ? :geocoder_google : :geocoder_here
+    cache_hits = 0
+    countable_requests = 0
+    from.upto(to).each do |date|
+      success = usage_metrics.get(geocoder_key, :success_responses, date)
```
Comment (juanignaciosl) [inline app/helpers/data_services_metrics_helper.rb]: Instead of `get` + addition with `nil` check you could do an straightforward addition by converting `get` return value to integer with `to_i`, which is nil-safe (`nil.to_i == 0`). Minor, though.

File: app/helpers/data_services_metrics_helper.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      countable_requests += success unless success.nil?
+      empty = usage_metrics.get(geocoder_key, :empty_responses, date)
+      countable_requests += empty unless empty.nil?
+      hit = usage_metrics.get(:geocoder_cache, :success_responses, date)
+      cache_hits += hit unless hit.nil?
+    end
+    countable_requests + cache_hits
```
Comment (juanignaciosl) [inline app/helpers/data_services_metrics_helper.rb]: How come are you using two different accumulators if you add both in the end?

Comment (ethervoid) [inline app/helpers/data_services_metrics_helper.rb]: Because I want to differenciate cache from countable_request for clarity and future changes

File: app/helpers/data_services_metrics_helper.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# encoding: utf-8
+
+require_relative '../../services/dataservices-metrics/lib/geocoder_usage_metrics'
+require_relative '../../services/dataservices-metrics/lib/routing_usage_metrics'
+
+module DataServicesMetricsHelper
```
Comment (juanignaciosl) [inline app/helpers/data_services_metrics_helper.rb]: I'm not sure about merging this into the same helper. Those methods present really low cohesion, it makes me think that keeping `GeocoderMetricsHelper` and creating the new `RoutingMetricsHelper ` might make more sense.

I'm also thinking about the opposite: generalizing, `get_*_data` methods, because they look really similar, but maybe you come up with something too complicated.

Just give this a thought if you haven't already, whatever decision you come up with will be fine.

File: app/models/carto/organization.rb [hunk 5]
```diff
@@ -41,6 +41,12 @@ def get_new_system_geocoding_calls(options = {})
       get_organization_geocoding_data(self, date_from, date_to)
     end
 
+    def get_routing_calls(options = {})
+      date_to = (options[:to] ? options[:to].to_date : Date.today)
```
Comment (juanignaciosl) [inline app/models/carto/organization.rb]: Maybe a `to_and_from_dates_from_options` method returning both to avoid duplication? Not 100% sure either.

File: app/models/carto/user.rb [hunk 6]
```diff
@@ -239,6 +240,15 @@ def remaining_geocoding_quota(options = {})
     (remaining > 0 ? remaining : 0)
   end
 
+  def remaining_routing_quota(options = {})
+    if organization.present?
+      remaining = organization.routing_quota.to_i - organization.get_routing_calls(options)
```
Comment (juanignaciosl) [inline app/models/carto/user.rb]: This logic belongs to `Organization`.

File: app/helpers/data_services_metrics_helper.rb (hunk 4 above)
Comment (ethervoid) [inline app/helpers/data_services_metrics_helper.rb]: The reason to not have two helpers was to avoid multiples helpers for the same functionality. This way I can include just one helper in detriment of have a repeated or similar code

File: app/models/carto/organization.rb [hunk 7]
```diff
@@ -41,6 +41,12 @@ def get_new_system_geocoding_calls(options = {})
       get_organization_geocoding_data(self, date_from, date_to)
     end
 
+    def get_routing_calls(options = {})
```
Comment (juanignaciosl) [inline app/models/carto/organization.rb]: Those `options` hashes everywhere are cursed. I know you've replicated previous design, but maybe you could at least document allowed keys (`:to` and `:from`, afaik).

File: app/models/carto/organization.rb (hunk 5 above)
Comment (ethervoid) [inline app/models/carto/organization.rb]: I've always disliked the methods with multiple elements returned

File: app/models/carto/organization.rb (hunk 7 above)
Comment (ethervoid) [inline app/models/carto/organization.rb]: Sure, no problem

File: app/models/organization.rb [hunk 8]
```diff
@@ -40,14 +40,16 @@ class Organization < Sequel::Model
...
 
   DEFAULT_GEOCODING_QUOTA = 0
+  DEFAULT_ROUTING_QUOTA = 0
 
   def validate
     super
     validates_presence [:name, :quota_in_bytes, :seats]
     validates_unique   :name
-    validates_format   /\A[a-z0-9\-]+\z/, :name, message: 'must only contain lowercase letters, numbers & hyphens'
+    validates_format   (/\A[a-z0-9\-]+\z/), :name, message: 'must only contain lowercase letters, numbers & hyphens'
```
Comment (juanignaciosl) [inline app/models/organization.rb]: Are parenthesis needed or suggested by Hound? (just curiosity).

Comment (juanignaciosl) [inline app/models/organization.rb]: screen shot 2016-02-16 at 17 26 42

Hahaha, Hound-trolling found afterwards.

Comment (ethervoid) [inline app/models/organization.rb]: Hound...

Comment (ethervoid) [inline app/models/organization.rb]: I hate it...

File: app/models/user.rb [hunk 9]
```diff
@@ -835,6 +863,15 @@ def remaining_geocoding_quota
     (remaining > 0 ? remaining : 0)
   end
 
+  def remaining_routing_quota
+    if organization.present?
+      remaining = organization.routing_quota - organization.get_routing_calls
```
Comment (juanignaciosl) [inline app/models/user.rb]: Same here.

File: db/migrate/20160212170700_add_routing_quota_and_block_price_fields.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
+Sequel.migration do
+  up do
+    add_column :users, :routing_quota, :integer, null: false, default: 0
+    add_column :users, :routing_block_price, :integer
+    add_column :users, :soft_routing_limit, :boolean
```
Comment (juanignaciosl) [inline db/migrate/20160212170700_add_routing_quota_and_block_price_fields.rb]: Could you consider default values and not null constraint for those columns, please?

File: services/dataservices-metrics/lib/service_usage_metrics.rb [hunk 11]
```diff
@@ -5,23 +5,9 @@
...
-      :geocoder_internal,
-      :geocoder_here,
-      :geocoder_google,
-      :geocoder_cache
-    ]
-
-    def initialize(redis, username, orgname = nil)
+  class ServiceUsageMetrics
+
+    def initialize(username, orgname = nil, redis=$geocoder_metrics)
```
Comment (juanignaciosl) [inline services/dataservices-metrics/lib/service_usage_metrics.rb]: Since `ServiceUsageMetrics` looks generic, I wouldn't favor `$geocoder_metrics` over others.

File: services/dataservices-metrics/lib/service_usage_metrics.rb [hunk 12]
```diff
@@ -49,14 +35,14 @@ def get(service, metric, date = DateTime.current)
...
     end
 
-    private
+    protected
 
     def check_valid_data(service, metric, amount = 0)
-      raise ArgumentError.new('Invalid service') unless VALID_SERVICES.include?(service)
-      raise ArgumentError.new('Invalid metric') unless VALID_METRICS.include?(metric)
-      raise ArgumentError.new('Invalid geocoder metric amount') if !amount.nil? and amount < 0
+      raise NotImplementedError.new("You must implement check_valid_data in your metrics class.")
```
Comment (juanignaciosl) [inline services/dataservices-metrics/lib/service_usage_metrics.rb]: Oh, so java-ish :heart: :-)

File: app/controllers/admin/organization_users_controller.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/controllers/admin/organization_users_controller.rb]: The only usage of this method is at `extras_enabled?` :-\

File: app/helpers/data_services_metrics_helper.rb (hunk 3 above)
Comment (juanignaciosl) [inline app/helpers/data_services_metrics_helper.rb]: Oki

File: app/helpers/data_services_metrics_helper.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/helpers/data_services_metrics_helper.rb]: Well, let's keep it this way.

File: app/models/carto/organization.rb (hunk 5 above)
Comment (juanignaciosl) [inline app/models/carto/organization.rb]: Ok

**Comment by juanignaciosl** [comment]
My major concerns were that `true` returned value at a method and the nullable columns, since you've already answered about the first one, :+1: after whatever you decide with the migration. Let me know if you want a second CR or not.

File: app/controllers/admin/organization_users_controller.rb (hunk 1 above)
Comment (ethervoid) [inline app/controllers/admin/organization_users_controller.rb]: Yes, I'm puzzle too but I don't want to break something in the other side of the code jajaja

File: app/models/carto/user.rb (hunk 6 above)
Comment (ethervoid) [inline app/models/carto/user.rb]: Right, I'll change it

File: db/migrate/20160212170700_add_routing_quota_and_block_price_fields.rb (hunk 10 above)
Comment (ethervoid) [inline db/migrate/20160212170700_add_routing_quota_and_block_price_fields.rb]: I did but then some parts of the code breaks because have logic for it XD so I removed it

File: services/dataservices-metrics/lib/service_usage_metrics.rb (hunk 11 above)
Comment (ethervoid) [inline services/dataservices-metrics/lib/service_usage_metrics.rb]: We could call it $service_metrics hahaha

File: app/models/carto/user.rb (hunk 6 above)
Comment (ethervoid) [inline app/models/carto/user.rb]: Done

File: app/models/user.rb (hunk 9 above)
Comment (ethervoid) [inline app/models/user.rb]: Done

File: app/models/carto/account_type.rb [hunk 13]
```diff
@@ -35,5 +35,15 @@ def soft_geocoding_limit?(user)
         user[:soft_geocoding_limit]
       end
     end
+
+    def soft_here_isolines_limit?(user)
+      if user[:soft_here_isolines_limit].nil?
+        # Due is an expensive service and the tiler could make multiples request
```
Comment (rafatower) [inline app/models/carto/account_type.rb]: Don't hate me but there's a more "elegant" way of expressing the same idea. The typical ruby idiom to return a boolean is:
```ruby
# if true the user can request isolines past their quota
# set to false by default to avoid unexpected charges
def soft_here_isolines_limit?(user)
  !!user[:soft_here_isolines_limit]
end
```

Comment (ethervoid) [inline app/models/carto/account_type.rb]: I have to say that I dislike that kind of cleverish type of code but I'll add it

File: db/migrate/20160212170700_add_routing_quota_and_block_price_fields.rb [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+    add_column :user_creations, :soft_here_isolines_limit, :boolean, default: true
+  end
+
+  down do
+    drop_column :users, :here_isolines_quota
+    drop_column :users, :here_isolines_block_price
+    drop_column :users, :soft_here_isolines_limit
+    drop_column :organizations, :here_isolines_quota
+    drop_column :organizations, :here_isolines_block_price
+  end
```
Comment (rafatower) [inline db/migrate/20160212170700_add_routing_quota_and_block_price_fields.rb]: Is the `user_creations` column missing here?

File: services/dataservices-metrics/lib/geocoder_usage_metrics.rb [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+module CartoDB
+  # The purpose of this class is to encapsulate storage of usage metrics.
+  # This shall be used for billing, quota checking and metrics.
+  class GeocoderUsageMetrics < ServiceUsageMetrics
+
+    VALID_METRICS = [
+      :total_requests,
+      :failed_responses,
+      :success_responses,
+      :empty_responses,
```
Comment (rafatower) [inline services/dataservices-metrics/lib/geocoder_usage_metrics.rb]: I love trailing commas but to be compliant with our style guide...

Also probably you should `CONSTANT = [ :sym1, :sym2 ].freeze` as hound suggests

File: db/migrate/20160212170700_add_routing_quota_and_block_price_fields.rb (hunk 14 above)
Comment (ethervoid) [inline db/migrate/20160212170700_add_routing_quota_and_block_price_fields.rb]: Yes, good catch!

File: spec/requests/signup_controller_spec.rb [hunk 16]
```diff
@@ -278,6 +281,7 @@
           end
 
           it 'triggers user creation with organization' do
+            Cartodb::Central.stubs(:sync_data_with_cartodb_central?).returns(false)
```
Comment (rafatower) [inline spec/requests/signup_controller_spec.rb]: why is this needed now?
